src/codegen/compile.ml

```diff
@@ -7895,6 +7825,10 @@ current buffer position.
 Note [mutable stable values]
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+NOTE: Currently not yet supporting 64-bit addresses during serialization and deserialization, e.g.
+* When the serialization offset is beyond 32 bit
+* When the deserialization target is beyond 32-bit (and sharing is used in the serialization)
```

Comment:
the 32 bit serialization offset seems less of an issue to me, but if the deserializaion target is not some offset but a 64-bit heap address, might we have problem when deserialization starts to allocate beyond the 4GB limit? Or am I misunderstanding what `target` is? A scenario might be upgrading from 32-bit to 64-bit canister with nearly 4GB of stable variables. The doubling oof value size might mean we need more 4GB on deserialization.

Reply:
Thanks for mentioning. This can indeed be a problem. I am not sure what a good solution is. At least we should test for the overflow. My hope is/was that we anyways could need deserialize more than 2GB on the classical system. So, maybe not support more than that. But probably too pragmatic approach. Alternatively, we would neeed to redesign the Candid destabilization logic by using some (dynamically allocated) mapping table, similar to graph copying deserialization.
What do you think?

Reply:
Linked it to https://github.com/dfinity/motoko/pull/4225/files#r1596453075

Reply:
**Update**: I fixed this problem by using pointer compression for the addresses stored in the memo (the target object pointer as well as the pointer to the type hash blob). I tested it manually (by increasing heap start in the RTS). Unfortunately, an automated integration test exceeds the batch size limit of `drun` (and even the instruction limit on `dfx`).